refactor(argus): Migrate to the REST Client

k0machi opened the pull request:
This commit makes SCT use the new REST API based client to interact with Argus. The idea of the new client is to be stateless and to avoid sending unnecessary data (leading to overwrites on the web application side), while also simplifying the use of the API by the end user and adding ease of updates as before any update had to be synced with SCT due to it directly interacting with Scylla cluster used in argus, making it sensitive to any schema changes.

## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [x] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [x] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [x] All new and existing unit tests passed (CI)
- [x] I have updated the Readme/doc folder accordingly (if needed)

k0machi commented:
Still missing heartbeat implementation, also I see some unit tests failing

k0machi commented:
Heartbeat implementation ready, fixed issues found by unit tests, this [job](https://argus.scylladb.com/test/b0b22734-758e-47b4-b6ec-29e64d212d52/runs?additionalRuns[]=72e6cd12-b0b0-444b-b955-4d3f6de47813) should succeed and mirror the results of the current implementation, if that's the case - PR is ready and I'm removing `sdcm/argus_test_run.py`

fruch commented on sdcm/tester.py:
```diff
@@ -382,14 +404,37 @@ def argus_test_run(self):
+                return
+            fail_count = 0
+            while not stop_signal.is_set():
+                if fail_count > 5:
```
Is that what was the logic before ?

Why would we stop trying ? cause argus might be down ?

Anyhow what information do we get out of those heartbeats ? what do we loose if well take them off ?

k0machi replied:
> Is that what was the logic before ?

Without the fail count

> Why would we stop trying ? cause argus might be down ?

In case argus is down / not responding to not spam logs with failed heartbeats.

> Anyhow what information do we get out of those heartbeats ? what do we loose if well take them off ?

Heartbeat timestamp is presented to the user when test is running, absence of it could indicate a crash or another problem.

k0machi replied:
> Without the fail count

To elaborate, it didn't have any exception handling, so the thread would exit on the first heartbeat failure.

fruch replied:
I'm not sure anyone actually uses that information as part of debugging jobs, most people are unaware of it.

From my POV we can remove it, it doesn't yet add too much value (in less we clearly state to the user that the job is "stuck", which we don't now)

k0machi replied:
> From my POV we can remove it, it doesn't yet add too much value (in less we clearly state to the user that the job is "stuck", which we don't now)

I think we should raise this question with the people before we remove it, and for improvement we could make frontend display a warning when a job heartbeat becomes too outdated - say 5 minutes. 

Additionally, there's another improvement in progress right now to clean up stuck jobs in argus that will rely on this field, so we can't just remove it.

fruch replied:
o.k. as a user it's a bit meaningless at least to me, that's why I was asking...

but if there other things going to be building ontop of it, that'sf fine...

k0machi commented on sdcm/cluster.py:
```diff
@@ -230,7 +229,7 @@ def __init__(self, name, parent_cluster, ssh_login_info=None, base_logdir=None,
         self.dc_idx = dc_idx
 
-        self.argus_resource: CloudResource | None = None
+        self.argus_resource = False
```
So this value was previously used when terminating a resource to decide if it needs to be terminated or not, since we don't really have state anymore we could just remove it if we're okay with an API exception in the log when the client tries to delete a non-existent resource.

fruch replied:
you'll have to explain it again, this flag now tell if we create the resource or not ? 
what are we trying to guard with it and why ? 

the only place I see it's being used is in `update_shards_in_argus`, which if we got there it's perfectly safe to assume we created the resource, anyhow we should be failing the test at that point one way or the other.

k0machi commented on sdcm/tester.py:
```diff
@@ -361,14 +368,29 @@ def __init__(self, *args, **kwargs):  # pylint: disable=too-many-statements,too-
+
     @cached_property
-    def argus_test_run(self):
+    def argus_connection(self):
```
Will do, think `argus_client` works better.

k0machi replied:
Done

fgelcer reviewed (approved):
LGTM, but shouldn't all the `argus` methos inside `tester.py` be moved to a centralized Argus file?

fruch commented on sdcm/cluster.py:
```diff
@@ -302,41 +301,57 @@ def init(self) -> None:
             LOGGER.error("Encountered an unhandled exception while interacting with Argus", exc_info=True)
 
-    def argus_resource_set_shards(self) -> CloudResource | None:
+    def argus_resource_set_shards(self) -> bool:
```
why this need to return boolean, the code doesn't check it.

fruch replied:
also rename it to `argus_set_shards`, `argus_resource` doesn't exists anymore

k0machi replied:
Done

fruch commented on sdcm/cluster.py:
```diff
@@ -302,41 +301,57 @@ def init(self) -> None:
+            LOGGER.error("Error creating argus client", exc_info=True)
+        return None
+
     def _init_argus_resource(self):
```
might consider rename this one as well, this doesn't init the resource
it's indicating start of a test

k0machi replied:
Done

fruch commented on sdcm/tester.py:
```diff
@@ -382,14 +404,37 @@ def argus_test_run(self):
             return unittest.mock.MagicMock()
 
+    def start_argus_heartbeat_thread(self) -> threading.Event:
+        def send_argus_heartbeat(connection: ArgusSCTClient, stop_signal: threading.Event):
```
again, connection or client, pick one...

k0machi replied:
Done

fruch commented on sdcm/cluster.py:
```diff
@@ -302,41 +301,57 @@ def init(self) -> None:
         self._init_argus_resource()
 
+    @cached_property
+    def argus_client(self) -> ArgusSCTClient | None:
```
you might consider moving this to `TestConfig` class, since we are creating this client multiple times
across different files/classes, testet.py / nemesis.py / common.py

k0machi replied:
Would it be easily accessible across modules, especially in `sct.py`? Maybe Fabio's idea with a central module would work better here?

fruch replied:
TestConfig is central, and unique across (Singleton) the whole SCT so that should fit the pattern you are looking for. creating a new Singleton also doable, would be the same from POV.

k0machi replied:
I think with current pattern we may be able to get away with just static methods with a cached class attribute containing the credentials (or client), so that may work even better

fruch commented on requirements.in:
```diff
@@ -31,7 +31,7 @@ mysql-connector-python==8.0.26
 python-jenkins==1.7.0
 ssh2-python==1.0.0
-argus-alm==0.8.1
+argus-alm==0.10.0
```
we must change version for this switch ?
since backporting this would be a bit problematic, since we'll need to rebuild hydra image for each branch (since there are other changes to hydra image which wasn't backported, and we don't know if not having them would break SCT or not)

k0machi replied:
I think 0.8.1 might be missing some fixes and improvements to the client, so might be annoying to not update

fruch replied:
if we don't have to have those improvement for release branches, I'm would rather not touch them.

For now let do this change on master, and we'll wait for a week before considering backports

k0machi replied:
Okay

fruch commented on sdcm/keystore.py:
```diff
@@ -98,6 +98,9 @@ def get_azure_credentials(self):
         return self.get_json("argusdb_config_v2.json")
 
+    def get_argus_rest_credentials(self):
+        return self.get_json("argus_rest_api.json")
```
maybe the file name should be `argus_rest_credentials.json` as the function name ? to avoid confusing ?

k0machi replied:
Done

soyacz commented on sdcm/utils/common.py:
```diff
@@ -772,18 +774,19 @@ def clean_instances_aws(tags_dict, dry_run=False):
+                try:
+                    argus_client.terminate_resource(name=name, reason="clean-resources: Graceful Termination")
+                except ArgusClientError as exc:
+                    if len(exc.args) >= 3 and exc.args[2] == 404:
```
Consider making ArgusClientError be more user-friendly. Using `exc.args >=3 and ex.args[2] == 404` is mysterious. Such things are unmaintainable.
e.g create new excetpion like "ArgusClientResourceDoesNotExistError" and throw it in client accordingly.

k0machi replied:
Will do in 0.10.1

soyacz commented on sdcm/utils/common.py:
```diff
@@ -756,11 +758,11 @@ def get_instances(region):
     aws_instances = list_instances_aws(tags_dict=tags_dict, group_as_region=True)
-    from sdcm.argus_test_run import ArgusTestRun  # pylint: disable=import-outside-toplevel
-    argus_run = ArgusTestRun.get()
+    creds = KeyStore().get_argus_rest_credentials()
+    argus_client = ArgusSCTClient(run_id=tags_dict["TestId"], auth_token=creds["token"], base_url=creds["baseUrl"])
```
these lines are repeating in multiple places. Can you hide it in some function? Only `test_id` looks customizable here.

k0machi replied:
Done

soyacz commented on sdcm/utils/common.py:
```diff
@@ -1220,23 +1223,27 @@ def clean_instances_gce(tags_dict, dry_run=False):
-                                                             reason="clean-resources: Graceful Termination")
-                argus_run.save()
             res = instance.destroy()
+            try:
+                client.terminate_resource(name=instance.name, reason="clean-resources: Graceful Termination")
+            except ArgusClientError as exc:
+                if len(exc.args) >= 3 and exc.args[2] == 404:
+                    LOGGER.warning("%s doesn't exist in Argus", instance.name)
+                else:
+                    LOGGER.error("Failure to communicate resource deletion to Argus", exc_info=True)
```
This is repeated, create function?

k0machi replied:
Done

soyacz reviewed (changes requested):
please fix repetition issues

fruch commented:
@k0machi few hydra docker images versions we merged since this PR was created.
please recreate the image when coming back to this one

k0machi commented:
> @k0machi few hydra docker images versions we merged since this PR was created. please recreate the image when coming back to this one

Rebased & Created new hydra version, also cleaned up repeated code. Decided against single module for now.

soyacz reviewed (approved):
LGTM

k0machi commented on sdcm/utils/common.py:
```diff
@@ -760,11 +761,10 @@ def get_instances(region):
 
 def clean_instances_aws(tags_dict, dry_run=False):
```
Yes, as argus only really needs instance name and run id to mark a cloud's instance as terminated, so no point in separate logic for now.

k0machi commented on sdcm/cluster.py:
```diff
@@ -306,43 +305,55 @@ def init(self) -> None:
 
-    def _init_argus_resource(self):
+    @cached_property
+    def argus_client(self) -> ArgusSCTClient | None:
```
Done

k0machi commented on sdcm/test_config.py:
```diff
@@ -278,3 +284,22 @@ def set_duration(cls, duration):
+    def init_argus_client(cls, params: dict):
+        if params.get("enable_argus"):
+            LOGGER.info("Initializing Argus connection...")
+            creds = KeyStore().get_argus_rest_credentials()
```
Yes, fixed.

k0machi commented on sdcm/tester.py:
```diff
@@ -364,35 +371,57 @@ def __init__(self, *args, **kwargs):  # pylint: disable=too-many-statements,too-
-                severity=Severity.WARNING,
-            ).publish_or_dump()
-            return unittest.mock.MagicMock()
+    def argus_client(self) -> unittest.mock.MagicMock | ArgusSCTClient:
```
Done

k0machi commented:
Unittest are failing with this:
```
ValueError: I/O operation on closed file.
```
Fastforwarded to the tip of the master because of that, could be an issue with removing `argus_client` messing with something in unit tests mocks.

fruch reviewed (approved):
LGTM

fgelcer commented:
@k0machi @fruch @soyacz @vponomaryov @roydahan , we need this backported to 5.0, 5.1, 5.2, 2022.1, 2022.2 and 2023.1 ASAP

fruch commented:
> @k0machi @fruch @soyacz @vponomaryov @roydahan , we need this backported to 5.0, 5.1, 5.2, 2022.1, 2022.2 and 2023.1 ASAP

it's already have it 5.2/2023.1, why we need it in the other branches ASAP ?

fgelcer commented:
> > @k0machi @fruch @soyacz @vponomaryov @roydahan , we need this backported to 5.0, 5.1, 5.2, 2022.1, 2022.2 and 2023.1 ASAP
> 
> it's already have it 5.2/2023.1, why we need it in the other branches ASAP ?

@k0machi can explain better, but new runs of these versions are not being reported in Argus without them (there are 4 commits in total to reach these active branches)

k0machi commented:
> > @k0machi @fruch @soyacz @vponomaryov @roydahan , we need this backported to 5.0, 5.1, 5.2, 2022.1, 2022.2 and 2023.1 ASAP
> 
> it's already have it 5.2/2023.1, why we need it in the other branches ASAP ? 

The argus table migrated recently, and those branches are still active and are reporting directly to the old one

fruch commented:
this PR is super problematic to backport:
1) it's change the docker image, and we can just slip the master image into those old branches - we might need to build new image as part of the backporting
2) it's not going to be a clean backport - since this PR touch lots of code paths

we'll need a PR for each branch - so it can be tested (at least unit test, and at least on job to verify)

fruch commented:
> > > > > @k0machi @fruch @soyacz @vponomaryov @roydahan , we need this backported to 5.0, 5.1, 5.2, 2022.1, 2022.2 and 2023.1 ASAP
> > > > 
> > > > 
> > > > it's already have it 5.2/2023.1, why we need it in the other branches ASAP ?
> > > 
> > > 
> > > The argus table migrated recently, and those branches are still active and are reporting directly to the old one
> > 
> > 
> > old branch reporting to old table, what changed that it's now urgent to backport ?
> 
> Fabio needs to see the new results since that branch is still considered active

and Argus isn't showing data from the older tables ????

k0machi commented:
> > > > > > @k0machi @fruch @soyacz @vponomaryov @roydahan , we need this backported to 5.0, 5.1, 5.2, 2022.1, 2022.2 and 2023.1 ASAP
> > > > > 
> > > > > 
> > > > > it's already have it 5.2/2023.1, why we need it in the other branches ASAP ?
> > > > 
> > > > 
> > > > The argus table migrated recently, and those branches are still active and are reporting directly to the old one
> > > 
> > > 
> > > old branch reporting to old table, what changed that it's now urgent to backport ?
> > 
> > Fabio needs to see the new results since that branch is still considered active
> 
> and Argus isn't showing data from the older tables ???? 

Yes, older tables are archived but there is currently no way for the user to switch to the old one

fruch commented:
> Yes, older tables are archived but there is currently no way for the user to switch to the old one

this is not very helpful schema migration process, are there any other options ? backporting all those change into multiple release branch would take a while.

k0machi commented:
> > Yes, older tables are archived but there is currently no way for the user to switch to the old one
> 
> this is not very helpful schema migration process, are there any other options ? backporting all those change into multiple release branch would take a while.

The previous table contained broken rows from a bug in 2021.1, it _seems_ to have gone away after cluster upgrade, but we've already migrated to a new table and started using it, so moving new runs back would also take some time. This backport would be nice as it means we can finally close off argus scylla instances to the outside world, since everything would be talking through an API instead.
